Use ginkgo v2 serial decorator

nunnatsa opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
### Add the ginkgo v2 Serial decorator
In functional tests, start replacing the `[Serial]` string within test spec descriptions, with the new ginkgo v2 `Serial`
decorators, by adding these decorators. Later we could remove the `[Serial]` strings from the spec descriptions.

### Use ginkgo v2 Serial decorator in several tests
Change hack/functests.sh to use the new ginkgo v2 Serial decorator:
1. drop the focus/skip for Serial, as ginkgo v2 uses the decorator.
2. Run only once, since ginkgo v2 uses the same algorithm as the old
version of this script: first it runs all the specks w/o the Serial
decorator in parallel, then it runs the tests with the Serial
decorator serialy.
3. rename the `--nodes` parameter with the new ginkgo v2 name `--procs`

Note: Several un-related files were deleted when running `make generate`.

**Special notes for your reviewer**:
For more details: https://github.com/nunnatsa/kubevirt/pull/new/use-ginkgo-v2-serial

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

oshoval commented on hack/functests.sh:
```diff
@@ -63,45 +63,27 @@ function functest() {
+fi
+
 if [ "$KUBEVIRT_E2E_PARALLEL" == "true" ]; then
     trap "_out/tests/junit-merger -o ${ARTIFACTS}/junit.functest.xml '${ARTIFACTS}/partial.*.xml'" EXIT
     parallel_test_args=""
```
you can remove now those two lines as well please

oshoval commented on tests/vm_test.go:
```diff
@@ -124,7 +124,7 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
 	})
 
-	Context("[Serial]A mutated VirtualMachine given", func() {
+	Context("[Serial]A mutated VirtualMachine given", Serial, Serial, func() {
```
no need twice ;]

oshoval reviewed (commented):
Thanks

oshoval commented on hack/functests.sh:
```diff
@@ -63,45 +63,24 @@ function functest() {
+if [ "$KUBEVIRT_E2E_PARALLEL" == "true" ]; then
     return_value=0
     set +e
-    functest --nodes=${KUBEVIRT_E2E_PARALLEL_NODES} ${parallel_test_args} ${KUBEVIRT_FUNC_TEST_GINKGO_ARGS}
+    functest --procs=${KUBEVIRT_E2E_PARALLEL_NODES} ${additional_test_args} ${KUBEVIRT_FUNC_TEST_GINKGO_ARGS}
     return_value="$?"
     set -e
     if [ "$return_value" -ne 0 ] && ! [ "$KUBEVIRT_E2E_RUN_ALL_SUITES" == "true" ]; then
         exit "$return_value"
     fi
```
can this be simplified now ?

nunnatsa commented:
/test pull-kubevirt-build-arm64

EdDev commented:
> Run only once, since ginkgo v2 uses the same algorithm as the old
version of this script: first it runs all the specks w/o the Serial
decorator in parallel, then it runs the tests with the Serial
decorator serialy.

Are you sure this is the "old" algorithm?
I remember it differently... that the serial ones run first and in case it fails, it does not continue.
If this changed, it is a major difference in how we run the tests and there may be multiple implications to this.

nunnatsa commented:
> Are you sure this is the "old" algorithm? I remember it differently... that the serial ones run first and in case it fails, it does not continue. If this changed, it is a major difference in how we run the tests and there may be multiple implications to this.

I'm pretty sure, yes, at least when running hack/functest.sh

oshoval commented:
> > Are you sure this is the "old" algorithm? I remember it differently... that the serial ones run first and in case it fails, it does not continue. If this changed, it is a major difference in how we run the tests and there may be multiple implications to this.
> 
> I'm pretty sure, yes, at least when running hack/functest.sh

It is indeed
https://github.com/kubevirt/kubevirt/pull/7620/files#diff-1315306f9c7e8e3278dfdf8ee00cefaa1ab1244deb08ae75411768fecb4a770aL94
Edy you can see here that the Serial is running after the parallel in the parallel block

oshoval commented:
Please see here
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7620/pull-kubevirt-e2e-kind-1.23-vgpu/1518198366016638976
The junit is missing, the junit is indeed empty, it seems the junit-merger ran on it
and the XSD format between the old reporter and the new are different, might be the reason

nunnatsa commented:
> Please see here https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7620/pull-kubevirt-e2e-kind-1.23-vgpu/1518198366016638976 The junit is missing, the junit is indeed empty, it seems the junit-merger ran on it and the XSD format between the old reporter and the new are different, might be the reason

I don't think this is the issue in the test you linked. There was a problem earlier when I removed the trap line from functest.sh. This line created the junit file. 

The problem here is, I think, something else:
```
/usr/local/bin/runner.sh: line 50: wait: pid 24 is not a child of this shell
```

Not sure where it come from.

oshoval commented:
> > Please see here https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7620/pull-kubevirt-e2e-kind-1.23-vgpu/1518198366016638976 The junit is missing, the junit is indeed empty, it seems the junit-merger ran on it and the XSD format between the old reporter and the new are different, might be the reason
> 
> I don't think this is the issue in the test you linked. There was a problem earlier when I removed the trap line from functest.sh. This line created the junit file.
> 
> The problem here is, I think, something else:
> 
> ```
> /usr/local/bin/runner.sh: line 50: wait: pid 24 is not a child of this shell
> ```
> 
> Not sure where it come from.

I would suggest to double check, as i saw those problems in the past when i tried to use the new reporter (which is what you are using), moreover it was a failure after you pushed the changed to the best of my knowledge,

lets first try to run it again simply ?

nunnatsa commented:
> I would suggest to double check, as i saw those problems in the past when i tried to use the new reporter (which is what you are using), moreover it was a failure after you pushed the changed to the best of my knowledge,
> 
> lets first try to run it again simply ?

I already re-ran. I'm not sure about the new reporter - I didn't change it. It was done before, I guess, when we moved to ginkgo v2.

The failed line is from here: https://github.com/kubevirt/project-infra/blob/13864c06908a530d812ea3b3e232b1e5be5c8371/images/bootstrap/runner.sh#L54

oshoval commented:
> > I would suggest to double check, as i saw those problems in the past when i tried to use the new reporter (which is what you are using), moreover it was a failure after you pushed the changed to the best of my knowledge,
> > lets first try to run it again simply ?
> 
> I already re-ran. I'm not sure about the new reporter - I didn't change it. It was done before, I guess, when we moved to ginkgo v2.

We still use the old reporter, even that we bumped to gingko 2.0
only this PR uses the new one

> 
> The failed line is from here: https://github.com/kubevirt/project-infra/blob/13864c06908a530d812ea3b3e232b1e5be5c8371/images/bootstrap/runner.sh#L54

Those are two different problem, that fact there isnt a junit report and the pid
if you open the artifacts file you will see it is wrong and empty

nunnatsa commented:
> > > I would suggest to double check, as i saw those problems in the past when i tried to use the new reporter (which is what you are using), moreover it was a failure after you pushed the changed to the best of my knowledge,
> > > lets first try to run it again simply ?
> > 
> > 
> > I already re-ran. I'm not sure about the new reporter - I didn't change it. It was done before, I guess, when we moved to ginkgo v2.
> 
> We still use the old reporter, even that we bumped to gingko 2.0 only this PR uses the new one
> 
> > The failed line is from here: https://github.com/kubevirt/project-infra/blob/13864c06908a530d812ea3b3e232b1e5be5c8371/images/bootstrap/runner.sh#L54
> 
> Those are two different problems, that fact there isnt a junit report and the pid if you open the artifacts file you will see it is wrong and empty

Found a problem with the reporter. But not the new format that I didn't change. The issue was that I restored the trap line to the wrong location in the file. Fixed now. Let's see if it will solve the issue.

oshoval commented:
we now have problems here with the junit merger
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7620/pull-kubevirt-e2e-k8s-1.23-sig-storage-cgroupsv2/1518208901458169856#1:build-log.txt%3A8993

Try please maybe to remove the trap, maybe you don't need to merge anymore, but just that it will know to create the right name directly by the test suite
`junitOutput = filepath.Join(flags.ArtifactsDir, fmt.Sprintf("partial.junit.functest.%d.xml", GinkgoParallelProcess()))`

nunnatsa commented:
/retest

EdDev commented on tests/performance/framework.go:
```diff
@@ -53,12 +56,27 @@ func init() {
 
-func FSIGDescribe(text string, body func()) bool {
-	return FDescribe("[sig-performance][Serial] "+text, body)
+func isIncludeSerial(body ...interface{}) bool {
```
This is needed because gingko considers multiple `Serial` items in the slice as an error? If this is the case, I tend to think it is a bug.

Anyway, I would prefer to fix the duplication and be done with it, over doing this nasty hack.
Is it possible?

nunnatsa replied:
This change is not directly related to this PR, but it's needed for this PR. The spec APIs were changed from v1 to v2, and so these custom envelopes (`SIGDescribe`, `FSIGDescribe`) need to be updated to match the API change of the `Describe` container. No hack here.

EdDev replied:
The comment is on `isIncludeSerial` addition, which checks inclusiveness of the `Serial` item. I am not talking about the signature change.

EdDev replied:
I peeked at the source-code of ginkgo. I do not really understand why we need this check at all.
It is suppose to be fine even if multiple identical `Serial` arguments are included in the args. What exactly did not work for you that you had to do this?

nunnatsa replied:
OK, so I don't think ginkgo cares about duplication here. I can drop this code, but I'll check ginkgo source code first, to be sure.

nunnatsa replied:
Fixed, thanks!

EdDev commented on tests/performance/framework.go:
```diff
@@ -53,12 +56,27 @@ func init() {
-	return FDescribe("[sig-performance][Serial] "+text, body)
+func isIncludeSerial(body ...interface{}) bool {
+	for _, item := range body {
+		if reflect.TypeOf(item) == serialType && item == Serial {
```
Will [this](https://pkg.go.dev/go/types#Identical) work here?

nunnatsa replied:
yes.

nunnatsa replied:
no, not really. The types package is not meant to work like that and won't help here.

nunnatsa replied:
Anyway, this code was removed as result of the previous comment

EdDev commented on tests/utils.go:
```diff
@@ -3527,7 +3527,8 @@ func UpdateKubeVirtConfigValue(kvConfig v1.KubeVirtConfiguration) *v1.KubeVirt {
 	suiteConfig, _ := GinkgoConfiguration()
-	if suiteConfig.ParallelTotal > 1 {
+	// ginkgo runs all the specs with the Serial decorator in process #1.
+	if (suiteConfig.ParallelTotal > 1) && (GinkgoParallelProcess() != 1) {
```
I am not clear why there is a need to check `suiteConfig.ParallelTotal > 1` now.
If it is equal to 1, won't `GinkgoParallelProcess()` return 1 as well?

nunnatsa replied:
I can't tell for sure. The ginkgo documentation does not deal with what is the process ID when running serially. I'm not sure we can count on it.

EdDev replied:
Fair enough. Thanks.

nunnatsa replied:
Fixed. Thanks!

EdDev reviewed (commented):
Nice one, I like this approach!

As soon as this gets merged, I would recommend sending a deprecation notification to the mailing list and setting a date for removal (lets say 3 months).

I have not yet gone through the script changes.

BTW, there is one side effect here: Unlike previously, failure in the parallel tests will not cancel the serial tests run. Some will see this as a good step and others as a bad one.
The thing that will really count is the CI load. If we'll notice a significant change in the load, there will be a need to handle it. At the moment, the only thing we can do is measure serial existing runs to get a sense of what is the potential cost.

EdDev commented on hack/functests.sh:
```diff
@@ -63,45 +63,18 @@ function functest() {
-    functest ${serial_test_args} ${KUBEVIRT_FUNC_TEST_GINKGO_ARGS}
-    exit "$return_value"
+if [ "$KUBEVIRT_E2E_PARALLEL" == "true" ]; then
+    trap "_out/tests/junit-merger -o ${ARTIFACTS}/junit.functest.xml '${ARTIFACTS}/partial.*.xml'" EXIT
```
As there is only one execution (and not two as before), is there still a need to perform the merges? Are there indeed individual files for each parallel run?
(I thought there are two: One for the parallel run and another for the serial one)

This was first raised [here](https://github.com/kubevirt/kubevirt/pull/7620#issuecomment-1107859634).

EdDev replied:
I think this is what I see in the artifacts as well [1].
There are always two "partial" junit files and the 3rd is the merge result.

On this PR result, I can see only a single one [2].

[1] https://gcsweb.ci.kubevirt.io/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7627/pull-kubevirt-e2e-k8s-1.23-sig-network/1518464819684970496/artifacts/

[2] https://gcsweb.ci.kubevirt.io/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7620/pull-kubevirt-e2e-k8s-1.21-sig-network/1518208895724556288/artifacts/

oshoval replied:
lets remove it and ask them to update their scripts ?
junit-merger is doing some conversion as well, so it is better to remove code (however it will still kept on unit-tests which need more work in order to remove it)
(same for the previous "[Serial]")

if you use junit-report and run the trap it will fail, because there will be duplicate sync sections
removing the trap will be a good preparation for adapting the ginkgo 2.0 built-in --junit report

EdDev replied:
> Unfortunately, we still need it, because other clients use these tests, and expect the partial.*.xml file name format. That can be easily fixed, using `--junit-report` parameter when running the tests.

What clients? I am unclear on why someone will depend on the intermediate artifacts.
I guess I am missing some context here, but my understanding that those files are just "partial" temporary artifacts from which the unified report is constructed. We are suppose to care about the end result, no?

EdDev commented:
> Note: Several un-related files were deleted when running `make generate`.

I have just ran it locally on the main branch and could not see any changes.
Could you rebase to main and re-check this?

nunnatsa commented:
> > Note: Several un-related files were deleted when running `make generate`.
> 
> I have just ran it locally on the main branch and could not see any changes. Could you rebase to main and re-check this?

This is after rebasing.

nunnatsa commented:
/retest

nunnatsa commented:
/retest

oshoval commented on hack/functests.sh:
```diff
@@ -60,48 +60,35 @@ function functest() {
     fi
 
-    _out/tests/ginkgo -timeout=3h -r -slow-spec-threshold=60s $@ _out/tests/tests.test -- -kubeconfig=${kubeconfig} -container-tag=${docker_tag} -container-tag-alt=${docker_tag_alt} -container-prefix=${functest_docker_prefix} -image-prefix-alt=${image_prefix_alt} -oc-path=${oc} -kubectl-path=${kubectl} -gocli-path=${gocli} -installed-namespace=${namespace} -previous-release-tag=${PREVIOUS_RELEASE ...
+    _out/tests/ginkgo -timeout=3h -r --junit-report="${ARTIFACTS}/junit.functest.xml" -slow-spec-threshold=60s $@ _out/tests/tests.test -- \
```
--junit-report is the new ginkgo 2.0 reporter, i would suggest to not mix those two features in the same PR
